samigo/samigo-services/src/test/org/sakaiproject/tool/assessment/services/GradingServiceTest.java

```diff
@@ -506,6 +505,24 @@ public void testInterestRateQuestion() throws Exception {
         Assert.assertEquals("866.2", result);
     }
 
+    @Test
+    public void testFactorial() throws Exception {
+        // This is the way a traditional mxParser factorial is written
+        String formula = "10/5-(5!)";
+        String result = gradingService.processFormulaIntoValue(formula, 1);
+        Assert.assertEquals("-118", result);
+
+        // This will replace the factorial with an iterated operator
+        formula = "10/5 + factorial(10) + 3/6";
+        result = gradingService.processFormulaIntoValue(formula, 1);
+        Assert.assertEquals("3628802.5", result);
+
+        // This is legacy from old, customer parser
+        formula = "FACTORIAL(4)";
+        result = gradingService.processFormulaIntoValue(formula, 1);
+        Assert.assertEquals("24", result);
+    }
```

Comment:
Perhaps assert that (-5)! gives NaN (I'm sure anyone publishing a math library with a factorial expression wouldn't let negative values trigger a S/O, but you can never be too safe)

Reply:
ok done